Refresh query when parameters update

gabrieldutra opened the pull request:
## What type of PR is this? (check all applicable)
<!-- Please leave only what's applicable -->
- [x] Feature

## Description
This is a work on #2316.

So far, I've added a touch state to Query Parameters and it now refreshes query results whenever a parameter is modified (debounce already added). Next step will be to add an "Apply" button for input fields while refining UX.

- [x]  Refresh query when parameters are modified
- [x] Add "Apply" button for input fields and let auto-refresh for the rest (Dropdowns, Time pickers)
- [x] Remove button from #3752

## Related Tickets & Documents
#2316 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
Not yet :sleeping:

gabrieldutra commented:
Thanks for the comments @arikfr, I've imagined this would be applied to other places where Parameters is used. I thought about doing as you mentioned before, the main reason I went with more logic in the `view` was thinking about the need of an "Apply All" functionality in the future (e.g: when clicking Execute Query button). I though about 2 ways to make it work:
- Having an exposed method in `Parameters` component that would force update, so I could use refs and when needed just run `parametersComponent.applyAll()` - however, I'm not so sure how to handle it with both Angular and React coexisting
- Lift state up - things end up in the `view` :laughing: (that's why I had changed `getQueryResults`)

Well, I agree that the second one is not ideal as we would repeat a lot of logic across components, so I'm moving to 1. Gonna make it work first without this functionality, then find a way to handle it. In any case, this is sort of what I had in mind.

Regarding debounce, NP, I'm more using it for testing, later we figure out where it should be.

arikfr commented:
@gabrieldutra I don't think the component need to expose a method, but rather it should use the same data model object that the parent component will use.

Also, not sure that executing a query should apply the new inputs. It's probably simpler to have the parent components see the old value until it's applied by the input component. This should make implementation simpler and makes sense UX wise.

gabrieldutra commented:
> Also, not sure that executing a query should apply the new inputs. It's probably simpler to have the parent components see the old value until it's applied by the input component. This should make implementation simpler and makes sense UX wise.

I thought there could be a case of pressing, for example, `Ctrl+Enter` all inputs would be applied. In any case, I'm with: "It's probably simpler to have the parent components see the old value until it's applied by the input component".

For the sake of moving forward, I've added the "Apply Button" and played a little bit with it:

**Current implementation**
![width-fixed](https://user-images.githubusercontent.com/3356951/56906008-cd540980-6a77-11e9-917f-4de97e86d71b.gif)

**With auto width** - it's kind of cool, but there's the issue of wrapping as you increase the final width
![width-auto](https://user-images.githubusercontent.com/3356951/56906079-ebba0500-6a77-11e9-9da4-eca0e987c8d6.gif)


There was an issue with Spinners in Number Inputs and the parameters, so I've removed them. LMK if you prefer a different solution :slightly_smiling_face: 

![number-input](https://user-images.githubusercontent.com/3356951/56907872-9122a800-6a7b-11e9-9749-4d996b483846.gif)

@arikfr, @ranbena suggestions for UX/UI here?

ranbena commented:
@gabrieldutra these are very cool suggestions.
* I would rather avoid a button within the input element cause it may cause unforeseen problems.
* Should there be a cancel button along with the apply? 

My tendency is to do something similar to the UI in param editor:
![Screen Shot 2019-04-30 at 22 24 38](https://user-images.githubusercontent.com/486954/56987689-c839ac00-6b96-11e9-96e8-30431872b0f7.png)

So perhaps something like this:

![1](https://user-images.githubusercontent.com/486954/56987949-80675480-6b97-11e9-9388-b230f8286563.jpg)

That when clicked turns to this:

![2](https://user-images.githubusercontent.com/486954/56987955-85c49f00-6b97-11e9-83d0-c37ad0b5fce1.jpg)

arikfr commented:
@ranbena there is no need for edit button, as this is an input field... Having the buttons outside of the control does indeed solve some issues, but is it clear what it will do?

Although I wonder if we should implement this UI:

![image](https://user-images.githubusercontent.com/71468/56988221-2d41d180-6b98-11e9-8f91-9ed692c5a29f.png)

(including static label for value until edited) 🤔 

I'll sleep on it.

@kravets-levko @rauchy wdyt?

ranbena commented:
> @ranbena there is no need for edit button, as this is an input field... Having the buttons outside of the control does indeed solve some issues, but is it clear what it will do?

I meant for this "edit" button to swap the "cog" (should keep it cog tho), and when `input.onChange` to swoosh the apply/cancel buttons in over it.

arikfr commented:
> I meant for this "edit" button to swap the "cog" (should keep it cog tho), and when input.onChange to swoosh the apply/cancel buttons in over it.

Oh, I see now. Well, the cog only visible for the query editor in edit mode, so not sure if I would bother with moving it or changing it.

My concern that it won't be intuitive what this confirm/cancel button do. But it seems like the simplest solution for now, so maybe worth starting with it and iterating from there.

@gabrieldutra wdyt?

rauchy commented:
I still find the word "Apply" more descriptive and easy to understand than an icon.

gabrieldutra commented:
> I would rather avoid a button within the input element cause it may cause unforeseen problems.

I did validate this for different resolutions, but I share the same concern. 

> Should there be a cancel button along with the apply?

If the button will be added in the "same place" as the input (not in a popover) I think it's better not to have a cancel button. Seems simpler without it + I don't think it adds much in this case.

Regarding Static label (or Popover): the idea is interesting, but it adds a new step to the process. I mean, when editing 1 parameter it's fine, but when editing 5 this may get annoying. Also, it wouldn't be possible to use only the keyboard when editing more than 1 parameter.

I agree with @rauchy that the word "Apply" is more intuitive. Perhaps explore it without it being within the input?

arikfr commented:
My thoughts after reviewing this again and playing with the current version:

* I agree that `Apply` is more intuitive. But placing it outside of the Input will most likely make it less intuitive, so I feel that this is not worth exploring.
* Auto width doesn't feel necessary, because the input is large enough as it is even after adding the button.

It seems that the only issue are the spinners for numbers -- I wasn't that attached to them, so don't mind them going away. If we get feedback they were useful, we can explore other options. But for now I think that your current implementation is good 👌 

Let's move on ⏩ and get this implemented :)

gabrieldutra commented:
I'll have to work on a different approach to it anyway, following @ranbena's comment, I've just tested it on Firefox:

![firefox-apply](https://user-images.githubusercontent.com/3356951/57109427-cfce9180-6d0b-11e9-96e9-f414286f3885.gif)

Remove the border from the input and put it in a `div` that wraps both the input and the button should be more reliable and give the same effect. I think I'll leave as is until I finish the core functionality, just in case other ideas come :)

arikfr commented:
>  I think I'll leave as is until I finish the core functionality, just in case other ideas come :)

Good call. Let's finish the functionality, add debounce where needed and do the FF fix in a follow up PR so we can release this already. While the bouncing inputs are annoying, it's still better than current behavior.

gabrieldutra commented:
I've updated this to include the Apply Button in the other places. @arikfr, is anything missing for this to be done in terms of functionality? If not, this can be open for review :slightly_smiling_face: 

The best place to put debounce seemed to be the `Parameters` component in the `onChange` method. Ideally this should debounce the parameter's `setValue` as well, but this will be easier when we migrate the `Parameters` component (by using states and so).

arikfr commented:
I think you can open this to review already. Let's finalize this before moving on to implementing the better rendering.

Two issues I did notice when testing it:

- [ ] When you hit "Cmd + Enter" inside an input dialog it will execute the query twice (because both the page & input dialog catch the Enter key event).
- [ ] DateTime Parameter: it should apply only when you click on `Ok` (or close the dialog?), but currently it applies immediately when picking some date. This is specially an issue with date/time input, because it doesn't give you the chance to pick the time.

gabrieldutra commented:
> When you hit "Cmd + Enter" inside an input dialog it will execute the query twice (because both the page & input dialog catch the Enter key event).

I've excluded both Meta (cmd) and Ctrl key, if someone can validate this for MacOS :).

> DateTime Parameter: it should apply only when you click on Ok (or close the dialog?), but currently it applies immediately when picking some date. This is specially an issue with date/time input, because it doesn't give you the chance to pick the time.

Antd's `DatePickers` don't have Ok buttons while `DateTimePickers` do, I've updated both `DateTime` and `DateTimeRange` to update values when you close the dialog (it was the one with apparently less issues). I used some hooks to manage state in the component :slightly_smiling_face:. There's an `onOk` callback which could be an alternative, but it keeps the selected values when you close the dialog regardless of the ok button, so it could be confusing.

gabrieldutra commented on client/app/components/parameters.html:
```diff
@@ -18,14 +18,6 @@
-    title="Refresh Dataset"
-  >
-    <span class="zmdi zmdi-play"></span>
-  </button>
```
@rauchy :)

ranbena commented on client/app/components/ParameterValueInput.less:
```diff
@@ -0,0 +1,19 @@
+  width: 195px !important;
+
+  &--apply-button {
+    padding-right: 60px !important;
```
To make it equal the 4px from the right edge.

```suggestion
    padding-right: 58px !important;
```

ranbena commented:
> It seems that the only issue are the spinners for numbers -- I wasn't that attached to them, so don't mind them going away. If we get feedback they were useful, we can explore other options. But for now I think that your current implementation is good 👌

I'm not sold on it but we can avoid the functionality regression by insetting the apply button (and persisting spinner opacity) for number inputs:

<img width="234" alt="Screen Shot 2019-05-09 at 17 07 18" src="https://user-images.githubusercontent.com/486954/57492862-c307eb80-727f-11e9-8e0d-757bda42a1ca.png">

ranbena commented on client/app/components/ParameterValueInput.less:
```diff
@@ -0,0 +1,19 @@
+@import '~antd/lib/input-number/style/index'; // for ant @vars
+
+.parameter-apply-button {
+  margin-left: -55px;
```
This is the cause of the FF display issue.
It's better to take the button "out of the flow" so it doesn't affect neighboring elements.

```css
parameter-value-input {
  position: relative;
  display: inline-block;
}

.parameter-apply-button {
  width: 50px;
  position: absolute !important;
  right: 4px;
  top: 6px; /* 5px after button height set to 25px */
}
```

ranbena commented on client/app/components/ParameterValueInput.less:
```diff
@@ -1,19 +1,50 @@
+    height: 25px !important;
+    width: 50px;
+    top: 5px;
+    right: 26px;
```
😬
```css
right: 27px;
```

gabrieldutra commented:
@ranbena, I've updated the styling after considering your comments :). What did you base `right: 4px;` in?

![apply-button-4px](https://user-images.githubusercontent.com/3356951/57541793-38c69280-7326-11e9-92dc-1f8d4e80249b.png)

Should it be 5px instead to fit top & bottom of Apply Button's position?

> When text is long and starting to edit, the edited area is suddenly out of bounds, so user has to stop and scroll it into view to keep editing..

This is sort of an edge case if you consider that when you go on typing it automatically goes to the caret position, I think if there is an easy solution we should do. I'll see what I find for this.

> Since all input param types now apply only on selection / apply click - no reason to debounce at all, right?

I think the idea is to have some time to apply more parameters without executing the query multiple times (although the 1 sec I let for tests seems not to be enough ^^)

ranbena commented:
> @ranbena, I've updated the styling after considering your comments :). What did you base `right: 4px;` in?

Looking good. 4px comes from the button's distance from top and bottom. For visual consistency it should be 4px from all sides.

> Should it be 5px instead to fit top & bottom of Apply Button's position?

With absolute positioning it should be `top: 5px; right: 5px` 👍 

> > When text is long and starting to edit, the edited area is suddenly out of bounds, so user has to stop and scroll it into view to keep editing..
> 
> This is sort of an edge case if you consider that when you go on typing it automatically goes to the caret position, I think if there is an easy solution we should do. I'll see what I find for this.

👍 

> > Since all input param types now apply only on selection / apply click - no reason to debounce at all, right?
> 
> I think the idea is to have some time to apply more parameters without executing the query multiple times (although the 1 sec I let for tests seems not to be enough ^^)

Why is it needed though? Doesn't the previous execution get cancelled on re-execution?
The debounce has a bad UX effect in which you click "apply" and nothing happens for a sec. It's annoying and we should avoid it.

ranbena commented:
> This is sort of an edge case if you consider that when you go on typing it automatically goes to the caret position, I think if there is an easy solution we should do. I'll see what I find for this.

@gabrieldutra just noticed that if you keep typing, Chrome will automatically scroll to caret. So can leave as is.

gabrieldutra commented:
> Why is it needed though? Doesn't the previous execution get cancelled on re-execution?
The debounce has a bad UX effect in which you click "apply" and nothing happens for a sec. It's annoying and we should avoid it.

Totally agreed, I'm not really sure if the previous get cancelled, for MySQL iirc it doesn't work ^^. I guess in such cases it would queue the query executions (which can be an issue for slow queries). For the sake of testing I've removed debounce:

![execution-queued](https://user-images.githubusercontent.com/3356951/57545030-74198f00-732f-11e9-9b9e-47e890b0744e.gif)
(BTW, interesting way to test slow queries ^^)

It seems that: - when there's one worker available, the triggered execution goes directly to it; - when the execution is queued (not running yet), it is replaced by the new one(?)

ranbena commented:
> Totally agreed, I'm not really sure if the previous get cancelled, for MySQL iirc it doesn't work ^^. I guess in such cases it would queue the query executions (which can be an issue for slow queries). For the sake of testing I've removed debounce:

Well I can tell you the user experience now is soooo much better :)
> (BTW, interesting way to test slow queries ^^)

Nice!

> It seems that: - when there's one worker available, the triggered execution goes directly to it; - when the execution is queued (not running yet), it is replaced by the new one(?)

@arikfr how does it work?

arikfr commented:
> It seems that: - when there's one worker available, the triggered execution goes directly to it; - when the execution is queued (not running yet), it is replaced by the new one(?)

We deduplicate executions of the same query multiple times, i.e. if the query already executes and you execute it again, it will "join" the previous execution. So in the above case, because you didn't change the values, it was all one execution.

Indeed the idea here was to avoid multiple executions when you want to update several parameters. But thinking about it, I see that debounce is not enough -- what needs to happen is more complex: if you start editing another parameter, we should wait with execution.

Let's roll out without debounce, and improve this based on feedback from users and our experience using this.

arikfr commented:
Is this ready in terms of implementation?

ranbena commented:
A sidenote on debounce and UX:
The problem is not the debounce itself, but the delay in "responsiveness". There must be a visual indication that the "apply" click has been registered and acted upon - even if in actuality it's idle.

Similar to the dashboard [save debounce implementation](https://github.com/getredash/redash/blob/master/client/app/pages/dashboards/dashboard.js#L90).

gabrieldutra commented:
> Is this ready in terms of implementation?

Considering the last comments, yes. I do have a concern about slow queries, but I think it's safe to go through "Let's roll out without debounce, and improve this based on feedback from users and our experience using this.".

> There must be a visual indication that the "apply" click has been registered and acted upon - even if in actuality it's idle.

This is sort of why I was keeping with 1 sec of debounce only, but definitively better without debounce or with a visual indication.

ranbena commented on client/app/components/ParameterValueInput.jsx:
```diff
@@ -132,25 +159,64 @@ export class ParameterValueInput extends React.Component {
-        onChange={onSelect}
-      />
+      <div
+        className={classNames('parameter-input-number', { 'parameter-input-number--apply-button': showApplyButton })}
```
When it comes to element state, I tend to favor describing the state rather than it's impact (cause that usually changes as app progresses) so instead of classname `parameter-input-number--apply-button` I'd go with `input-dirty`. And if it's a bool, I'd even ditch classnames and do a `data-dirty="true"` attribute on the element.

For your consideration 👨‍💻

ranbena replied:
Oh, this is just to have the apply button on/off 🤦‍♂️

ranbena replied:
Hmm so I would have these as two separate attributes actually.

Where does `$ctrl.applyButton` determined? When is it `false`?

ranbena replied:
Found it - hardcoded to `true`. Then perhaps this prop can be removed and code simplified?

gabrieldutra replied:
> When is it `false`?

`ParameterMappingInput` :slightly_smiling_face: (e.g: in the Add Widget Dialog). Also I thought it could be interesting to support it without the Apply Button.

Regarding the css, yes, it's to turn on/off the appearing of the apply button on the input. I tried to follow [BEM naming convention](http://getbem.com/naming/) on that, but yes, I can use a data attribute instead. It will actually look better than this `classNames` syntax.

gabrieldutra replied:
BTW, it was Cypress that made me realize that the Apply Button would affect the Add Widget Dialog :grin:

arikfr replied:
> I tried to follow BEM naming convention on that

Maybe we should have a conversation on whether we want to keep using BEM.

gabrieldutra replied:
Updated :+1:

> Maybe we should have a conversation on whether we want to keep using BEM.

This was actually the only case where I was applying BEM (modifiers like `--apply-button`). But I preferred as @ranbena suggested `data-dirty`. I think BEM is sort of deprecated for Redash already, just need to update the docs if that's really the case.

ranbena commented:
Another sidenote - the reason I prefer an icon to textual label in this case ("apply") is that some day we'll incorporate i18n, and this won't do 🔮

arikfr commented:
> Another sidenote - the reason I prefer an icon to textual label in this case ("apply") is that some day we'll incorporate i18n, and this won't do 🔮

🤔 fair point. Let's keep it as is for now with Apply and revisit in the future. I feel like the whole UI for parameters is due for deeper UX work. One option is to have it in a non editable form when presenting and have a proper UI for editing. Here's an example from Intercom:

![image](https://user-images.githubusercontent.com/71468/57598504-2e8dd980-755c-11e9-8c59-aeff78bcb0d1.png)

And when editing:

![image](https://user-images.githubusercontent.com/71468/57598523-3fd6e600-755c-11e9-8cf4-3111fd7d1086.png)

---

About an alternative to debounce, we could have a dropdown next to the Apply. 

![](https://cl.ly/298cb611ad42/Screen%252520Recording%2525202019-05-13%252520at%25252008.56%252520AM.gif)

~Not sure it's appropriate considering the limited space we have for it.~ We can't do it with current UI, because it will work only with `<Input>`s but not dropdowns or dates. But this is something we can do once we switch to rendering a dedicated UI for editing.

arikfr commented:
Date Range parameters don't seem to update on dashboards. See example here: https://deploy-preview-3737--redash-preview.netlify.com/dashboard/legacy-parameters-dashboard?p_dateparam=2019-01-21&p_AlbumId=2001-01-20&p_time_frame.start=2019-01-03&p_time_frame.end=2019-01-11&p_param1=title1&p_date%20range.start=2019-05-24&p_date%20range.end=2019-05-27.

gabrieldutra commented:
> Date Range parameters don't seem to update on dashboards

Actually none of the parameters are updating in there, it seems that `debounce` had an extra job for the order of functions. I'll dig into it and try to come up with sth more reliable for this part.

gabrieldutra commented on client/app/components/ParameterValueInput.jsx:
```diff
@@ -181,15 +244,21 @@ export default function init(ngModule) {
       this.setValue = (value) => {
         this.param.setValue(value);
-        $scope.$applyAsync();
+        $scope.$apply();
```
@arikfr changing `$applyAsync` to `$apply` here did the trick for the parameters not working on dashboards. What happens is that dashboards depends on Url Parameters to be updated (and it's a `watch` that does it) and the refresh was being triggered before.

arikfr replied:
Luckily soon we will get rid of all these crap 😎

arikfr commented:
🤩
